### PR TITLE
Refactored codebase to basic serial C++ implementation

### PR DESCRIPTION
**Pull request by tom91136**
This is a purely serial C++ implementation that can be used as starting points for other ports.
The only abstraction are as follows:
 * `clover::par_ranged{1,2}` - the implementation for both 1d and 2d are simple for loops, calls to these two functions will be inlined by hand at a later commit
 * `clover::Buffer{1,2}D` - allocating 2d arrays via new[] is a bit awkward and probably has very poor locality, so was going for something like `auto xs = new T[sizeX * sizeY];` but then we need to hold on to the sizes for indexing (`xs[j*sizeX + i]`). It's all inside a struct now. Buffer1D doesn't have the same constraint but it is there for consistency.

It passes built-in, bm16, and bm16_short tests. 

This PR is ready for review now but there will be a follow-up commit where `clover::par_ranged{1,2}` gets inlined.

### CODE REVIEW

**Review by tomdeakin — CHANGES_REQUESTED** [review]
I'm concerned about both those abstractions, for some related but also different reasons.

The `clover::par_ranged{1,2}` abstraction hides the `for` loops which should exist in a serial C/C++ version. I expect this abstraction will be problematic when adding OpenMP directives, especially with the `collapse` clause which needs perfectly nested loops. Anything that the compiler might inline between the loops as a result of the abstraction will cause compilation errors. The lambda functions particular might cause that to happen. Additionally, lambda capture for OpenMP `target` directives sounds unintuitive and I expect it won't always be as robust as just using `for` loops and a loop body. I cannot see the advantage of using lambda functions here.

There is also the abstraction over the loop extents in `clover::Range{1,2}d`. This is used as a variable named `policy` but that is conflicting with what "policies" are in C++.

In short, we need to remove both these abstractions at this point and replace them with nested `for` loops.

The `clover::Buffer{1,2}D` abstraction might be OK. It might be a convenient way to deal with the initial mapping of the data when we come to OpenMP 4.5 (you can have `target enter data map(alloc: ...)` in the constructor). When using it in the kernel though we are going to need to know that `target` regions correctly capture scalars inside a `struct`. I think it should work, but as many of the dimensions are shared, we will store those `int` multiple times which h might create a lot of register pressure.

I think this we can see how it goes for this one, but be prepared in case we need to replace all the interfaces with the raw pointers in the future.

**Comment by tom91136** [comment]
The `par_ranged{1,2}d` will be going away (inlined) as planned. It was there because I was hoping clangd can help me with inlining, after wrestling with it last night I decided to do it by hand so it's gonna take some time. 
Since the repo is producing correct results with the lambdas now, I thought I'll at least leave a commit in place so in case some other Kokkos-like abstractions shows up, we can go from the lambdas to that a bit easier.

**Review by tomdeakin — COMMENTED** [review]
This looks good to me now, thanks!

**Review by tomdeakin — APPROVED** [review]
